[2단계 - 웹 자동차 경주] 후추(주찬민) 미션 제출합니다.

Combi153 opened the pull request:
안녕하세요 카일 😄

후추입니다 🤸‍♀️🤸

요즘은 기분 좋은 봄날의 연속이네요! 따뜻한 봄 잘 보내고 계신가요?

저는 지난 주말부터 화요일까지 감기몸살을 앓아 미션을 거의 진행하지 못했어요

그래서 2단계 미션 제출이 많이 늦어졌습니다 죄송해요 😭 

오늘이라도 제출할 수 있어 기쁘네요!

2단계도 잘 부탁드립니다 열심히 학습할게요!!!

### 지난 코멘트 답변
1단계 PR에 달아주신 코멘트에 답변과 질문을 달았어요! 제가 카일의 리뷰를 정확히 이해한 것인지 헷갈리는 부분도 있습니다 한번 확인해주신다면 감사할 것 같습니다
[지난 PR 링크](https://github.com/woowacourse/jwp-racingcar/pull/87)

### 질문
이번 미션을 진행하면서 크고 작은 궁금증들이 정말 많이 생겼어요. 그중 스스로 해결하기 힘들었던 부분을 질문 남깁니다!
1. Dao에서 쿼리를 통해 얻은 값을 `객체`에 담아 다른 계층에 넘겼습니다. 이때 이러한 객체를 무엇이라 이름 지어야 할 지 모르겠습니다. 계층 간 데이터 전송을 위한 객체이니 DTO라고 일단 이름을 지었습니다. 그러나 Controller에서 사용되는 DTO와 Dao에서 사용되는 DTO가 구분되지 않는 단점이 보입니다. 이름을 무엇으로 짓는 것이 적절할까요?

2. 위 질문에서 이어지는 질문입니다. 과연 Controller에서 사용되는 DTO와 Dao에서 사용되는 DTO가 구분되어야 할까요? 제 현재 코드에는 코드가 동일한 `CarDTO`와 `CarNamePositionDTO`가 존재합니다. 두 DTO는 사용처와 이름만 다릅니다. 이것에 대해 두 가지 생각이 듭니다.
- 계층 간 데이터 전송의 목적을 달성하는 데는 두 DTO 중 하나만 있어도 된다. 코드의 중복을 제거하자.
- 데이터 전송의 형태가 계층마다 독립적으로 변화할 수 있다. 변화에 대응하기 위해 두 개의 DTO를 만들자.
카일의 생각은 어떠신지 궁금합니다.

3. WebController에서 `@ExceptionHandler`를 사용해 예외 상황에 대한 응답을 보내도록  작성했습니다. 그런데 이러한 코드는 어떻게 테스트를 할 수 있는지 모르겠습니다. 흔히 가짜 객체를 만드는 방식을 택하는 것을 보았습니다만, 가짜 객체를 만들지 않고도 테스트가 가능한지 궁금합니다. 

4. 카일은 쉽게 해결되지 않는 궁금증을 마주했을 때 어떻게 하시나요? 구글링 실력이 부족한 탓인지, 구글링을 해도 쉽게 해소되지 않는 궁금증들이 있습니다. 이러한 궁금증은 어떤 방식으로 해소하시는 편이신지 궁금합니다!

Combi153 committed:
build: validation 의존성 추가

Combi153 committed:
feat: Controller에서 요청값에 대한 유효성 검사 추가

Combi153 committed:
feat: Service에 @Transactional 적용

Combi153 committed:
docs: 기능 요구 사항 추가 및 정리

Combi153 committed:
feat: GameIdDTO 추가

Combi153 committed:
feat: GameDao 테스트를 위한 기능 추가

- 행 개수 세기 기능
- 모든 데이터 삭제 기능

Combi153 committed:
feat: CarDao 테스트를 위한 기능 추가

- 행 개수 세기 기능
- 모든 데이터 삭제 기능
- 기존 insert 기능 반환값 설정

Combi153 committed:
feat: GameDao 모든 gameId 가져오는 기능 추가

Combi153 committed:
feat: CarDao에 한 게임의 우승자 이름 가져오는 기능 추가

Combi153 committed:
feat: CarDao에 한 게임의 모든 자동차 이름, 최종 위치 가져오는 기능 추가

Combi153 committed:
feat: SQL 데이터 삭제 시 설정 추가

- Car 테이블에서 foreign key game_id에 CASCADE 추가

Combi153 committed:
test: 각 테스트 실행 종료 시 테이블 삭제 기능 추가

Combi153 committed:
refactor: DTO 객체 이름 변경

- RequestDTO -> RacingStartDTO
- ResponseDTO -> RacingResultDTO

Combi153 committed:
style: 개행 삭제

Combi153 committed:
feat: 서비스에 게임 이력 조회를 위한 기능 추가

Combi153 committed:
feat: controller에 게임 이력 조회 기능 추가

Combi153 committed:
refactor: 랜덤 값 생성 의존관계 변경

Combi153 committed:
refactor: DTO에 equals, hashcode 메서드 삭제

Combi153 committed:
refactor: console application 출력 변경

- 플레이 중간 과정 출력 로직 제거
- 우승자와 참여자의 최종 위치 출력으로 변경
- 출력 메시지 enum 삭제
- OutputView 코드 수정

Combi153 committed:
refactor: 중복 코드 제거

- console application 로직이 web application의 코드를 사용하도록 수정
- OutputView 코드 수정
- Local Dao 추가

Combi153 committed:
refactor: Service 로직 수정

Combi153 committed:
refactor: 패키지 구조 변경

Combi153 committed:
refactor: console 관련 로직 수정

- InputView, OutputView 의존관계 삭제
- Application 객체 이름 변경

Combi153 committed:
refactor: 자동차 이름 예외 로직 수정

- 이름이 null 혹은 빈 값인 경우 예외

Combi153 committed:
feat: WebController 예외 처리 로직 추가

KIMSIYOUNG commented on src/main/java/racingcar/controller/ConsoleController.java:
```diff
@@ -0,0 +1,24 @@
+        this.racingCarService = racingCarService;
+    }
+
+    public void createGameAndPlay() {
```
게임을 플레이한다 정도로만 네이밍을 작성해도 될 것 같아요.
메소드 네이밍은 자신이 하고 있는 `기능` 또는 `역할` 을 드러내는 것이지, 내부 구현을 드러내는 것은 지양하면 좋을 것 같습니다.
1. createGame 은 게임을 진행하기 위해 내부적으로 수행하는 동작인 것 같아요.
2. 컨트롤러가 create한다는 개념을 알고 있다는 것도 낯서네요! (2줄에는 create 관련 로직이 없어요.)

Combi153 replied:
현재 컨트롤러의 네이밍이 `ConsoleController`, `WebController` 이기에 무엇을 play한다는 것인지 유추하기 힘들 것이라 판단했습니다. 그래서 createGame을 더했던 것 같습니다!

앞으로는 메서드 이름이 기능만 드러낼 수 있도록 신경쓰겠습니다 😄  

메서드 네이밍을 play로 수정하고 Controller의 이름도 CarRacingController 등으로 수정하겠습니다!

KIMSIYOUNG replied:
넵넵 좋은 방법인 것 같아요.

KIMSIYOUNG commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -18,8 +26,38 @@ public WebController(final RacingCarService racingCarService) {
+            errorMessageByFields.put(fieldName, errorMessage);
+        });
+
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(errorMessageByFields);
```
ResponseEntity는 badrequest 라는 정적메소드를 지원하고 있어요!

Combi153 replied:
알려주셔서 감사해요! 활용해보겠습니다 👍

KIMSIYOUNG commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -18,8 +26,38 @@ public WebController(final RacingCarService racingCarService) {
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<String> handleIllegalArgument(IllegalArgumentException exception) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(exception.getMessage());
```
이렇게 되는 경우 예외를 유실하게 됩니다. 로그에 대해서도 가볍게 학습해보고, 적용해보면 좋을 것 같아요.

Combi153 replied:
넵 로그를 적용해보겠습니다!

KIMSIYOUNG commented on src/main/java/racingcar/controller/WebController.java:
```diff
@@ -18,8 +26,38 @@ public WebController(final RacingCarService racingCarService) {
+        return ResponseEntity.ok(gameResults);
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
```
이러한 예외가 발생하는 케이스를 테스트하면 될 것 같은데, 어떤 부분이 고민이신지 조금 더 구체적으로 말씀해주시겠어요?

Combi153 replied:
챗 지피티에게 물어보고 궁금증을 해결했습니다! 👍  본래는 RestAssured를 통해 예외 테스트를 어떻게 작성해야 할지, 작성할 수는 있는 것인지 몰라서 여쭤본 것이었습니다.

KIMSIYOUNG replied:
아하 네네!

KIMSIYOUNG commented on src/main/java/racingcar/dao/LocalGameDao.java:
```diff
@@ -0,0 +1,62 @@
+
+public class LocalGameDao implements GameDao {
+
+    private final List<GameEntity> gameEntities = new ArrayList<>();
```
질문에 대한 이야기를 여기서 해보면 좋을 것 같아요.

1. Entity라는 네이밍으로 잘 구현해주신 것 같습니다. 찾아보시면 적절한 용어였다고 느끼실 것 같아요.
2. 이 부분은 프로그램의 크기나 요구사항, 발전방향 등에 따라 다를 것 같아요. 그렇기에 적절한 선택지가 무엇이다라는 답변은 드릴 수 없을 것 같습니다. 말씀하신대로 2가지 방법의 장단이 존재합니다. 이를 인지하고 있고 현재의 어플리케이션에서 무엇을 택할것인지를 결정하면 될 것 같습니다.

Combi153 replied:
넵 답변 감사드려요 😄
1. 익숙하지 않은 개념이라 찾아서 정보를 얻어도 긴가민가 한 것 같습니다. 앞으로 경험을 쌓으면서 제게 확신이 생기면 좋겠어요!
2. 아하 그렇군요. 역시 개발에는 정답이 없는 부분이 많은 것 같습니다. 답변 감사드려요 !!!

KIMSIYOUNG commented on src/main/java/racingcar/dto/RacingStartDTO.java:
```diff
@@ -0,0 +1,29 @@
+public class RacingStartDTO {
+
+    @NotBlank(message = "이름을 입력해야 합니다.")
+    private String names;
```
입력을 List로 받을 수도 있습니다. 또한 리스트안에 들어가는 값에 대해서도 검증할 수 있습니다.

Combi153 replied:
입력을 List로 받을 수 있다니 몰랐습니다. 한번 실험해보면서 방법을 익혀 보겠습니다!

실험도 해보고 검색도 해봤는데 입력을 바로 List로 받는 방법을 모르겠습니다. 방법을 알려주시면 감사하겠습니다 😢

KIMSIYOUNG commented on src/main/java/racingcar/dao/LocalCarDao.java:
```diff
@@ -0,0 +1,82 @@
+    @Override
+    public int insert(final String name, final int position, final Long gameId, final boolean isWin) {
+        carEntities.add(new CarEntity(name, position, gameId, isWin));
+        return 1;
```
구현한다면 제대로 구현해주시면 좋을 것 같습니다. 아이디가 항상 1이겠네요

Combi153 replied:
jdbcTemplate 의 update 메서드와 비슷하게 기능하도록 만들려고 했습니다. 따라서 변경된 행의 개수를 반환합니다. 여기서 1은 id가 아닌 변경된 행의 개수입니다. 의도를 드러낼 수 있도록 상수로 선언해 사용하겠습니다!

혹시 위와 같은 방식이 아닌, id를 반환해야 할까요?

KIMSIYOUNG replied:
음 제가 다른 분과 코드를 착각한 것 같아요.
일반적으로 저장이후에는 저장된 아이디 값을 반환하는데요!
후추는 업데이트된 갯수를 반환하니 그대로 사용하셔도 될 것 같습니다. 

저는 JdbcDao는 아이디를 반환하는데 여기선 업데이트 된 수를 반환하는 줄 알았어요.

KIMSIYOUNG commented on src/main/java/racingcar/dao/LocalCarDao.java:
```diff
@@ -0,0 +1,82 @@
+
+public class LocalCarDao implements CarDao {
+
+    private final List<CarEntity> carEntities = new ArrayList<>();
```
리스트보다는 맵이 조금 더 적절한 자료구조일 것 같습니다.

Combi153 replied:
동의합니다! Map<Long, CarEntity>로 자료구조를 변경하고, Key 값으로는 id를 둘 수 있도록 리팩터링하겠습니다.

KIMSIYOUNG commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -1,52 +1,54 @@
 import java.util.stream.Collectors;
 
 @Service
+@Transactional
```
트랜잭션에 대해서는 천천히 학습해보시면 좋을 것 같습니다. 

추가로 리드온리 옵션에 대해서는 이번에 같이 학습해보시면 좋을 것 같아요.

Combi153 replied:
넵 키워드 알려주셔서 감사합니다! 제가 이해한 바는 다음과 같습니다.

데이터베이스의 CRUD 중 R은 데이터를 읽는 연산입니다. 따라서 쿼리 실행 전후로 데이터에 변화가 일어나지 않습니다. 이러한 경우 `@Transactional(readOnly = true)`로 설정해주면 해당 메서드가 읽기 전용이라는 것을 명시하고, 해당 트랜잭션에 대해 최적화를 할 수 있습니다.

이런 내용을 바탕으로 리팩터링을 진행한다면, Service 클래스 전체에 `@Transactional` 어노테이션을 붙이지 않고, 메서드마다 어노테이션을 붙이고 설정을 하는 것이 합리적일 것 같습니다.

Service 클래스에서 CUD와 관련된 메서드는 `@Transactional`을, R과 관련된 메서드는 `@Transactional(readOnly = true)`로 설정하겠습니다.

KIMSIYOUNG replied:
잘 학습해주셨네요😀과정을 진행하면서 천천히 더 깊게 학습해보면 좋을 것 같습니다. (어떻게 원자성을 보장하는지, 어떻게 리드쿼리는 최적화가 가능한지 등)

KIMSIYOUNG commented on src/main/java/racingcar/service/RacingCarService.java:
```diff
@@ -69,17 +71,37 @@ private List<String> getWinners(final GameSystem gameSystem) {
                 .collect(Collectors.toList());
     }
+
+    public List<RacingResultDTO> showGameResults() {
```
DTO를 변환하는 로직이 서비스의 사실상 대부분을 차지하네요.

1. `DTO`변환의 책임은 누가 가져가는게 좋을까요?
2. 현재 콘솔컨트롤러와 웹컨트롤러가 우연히도 같은 응답을 주도록 요구사항이 들어왔네요. 요구사항이 달라져서 콘솔컨트롤러의 응답이 달라지면 어떻게 변경하실예정이신가요?

Combi153 replied:
1. 
- DTO는 계층 간 데이터 전송을 목적으로 하는 객체입니다.
- DTO 변환의 책임은 각 계층이 가져야 합니다. 계층을 Presentation(Controller), Business(Service), Persistence(DAO)로 구분해본다면, 현재 Business 계층에는 Service 객체가 유일하므로, Service 객체가 책임을 져야합니다.
- 프로그램이 커진다면 이러한 방식이 Service 객체에게 과도한 책임을 지울 수 있을 것 같아요. 따라서 DTO 변환을 책임지는 객체를 만드는 것이 어떤가 생각합니다!

2. 서비스를 인터페이스로 두고, 새로운 서비스를 만들어 대응하는 방법이 있을 것 같습니다. 또한 DTO 변환 객체가 달라진 응답을 처리하도록 하는 방법도 떠오릅니다. 하지만 이것이 맞는 방법인지 구체적으로 어떻게 구현해야 하는지는 모르겠습니다. 고민이 필요한 부분인 것 같아요.

KIMSIYOUNG replied:
1. 좋은 방법인 것 같습니다. DTO에 로직을 넣는 것은 지양되어야 하지만 최소한의 변환정도는 스스로 해도 무방하다는 생각이 들어요.
2. 넵넵 그 방법도 좋은 것 같습니다. 제가 이 질문을 통해 고민해보시길 바란건 아래의 두가지입니다. 고민해보시면 좋을 것 같아요.
    - 서비스를 공통으로 사용하고 있는데 이것은 적절한 추상화인가?
    - DTO변환의 로직을 서비스에서 가져가는 것과 컨트롤러에서 가져가는 것의 장단은 무엇일까?

KIMSIYOUNG reviewed (changes requested):
안녕하세요 후추!
2단계 잘 구현해주셨네요.
몇가지 코멘트 남겼으니 확인 부탁드립니다.
1단계 PR에도 의견 달아두었는데 답변이 덜 되었거나 궁금하신 내용 있으시다면 현재 MR에 같이 작성해주시면 좋을 것 같습니다.

Combi153 committed:
refactor: controller 메서드 이름 수정

Combi153 committed:
refactor: Application, Controller, Service 네이밍 수정

- CarRacing 을 드러내도록 수정
- CarRacingWebController 에서 ResponseEntity badRequest 정적 메서드 활용

Combi153 committed:
refactor: 매직넘버 상수 처리

Combi153 committed:
feat: 로그 기능 추가

Combi153 committed:
refactor: LocalDao 의 자료구조 Map으로 변경

Combi153 committed:
test: LocalDao test 코드 추가

Combi153 committed:
test: WebController 예외 test 코드 추가

Combi153 committed:
refactor: Service에 Transactional 선언 내용 수정

Combi153 committed:
refactor: DTO 패키지 구조 변경

Combi153 committed:
refactor: 모든 Entity 클래스 필드에 id 추가

Combi153 committed:
refactor: 메서드 이름 변경

Combi153 committed:
refactor: Service 로직 리팩터링 및 DTO 클래스 이름 변경

Combi153 committed:
refactor: DTO 패키지 구조 변경

Combi153 commented:
안녕하세요 카일 😄

후추입니다 🤸‍♀️🤸

지난 리뷰를 통해 공부를 하고, 코드를 여러군데 수정해보았습니다! 리뷰 감사드립니다 😃 

## 지난 코멘트 답변

1단계 PR에 달아주신 코멘트에 답변을 달았어요! 한번 확인해주신다면 감사할 것 같습니다
[DTO의 필드 타입](https://github.com/woowacourse/jwp-racingcar/pull/87#discussion_r1174336525)

## 질문

코멘트에 남긴 질문을 제외하고 특별한 질문은 없습니다.

리뷰 잘 부탁드립니다!!

KIMSIYOUNG commented on src/main/java/racingcar/controller/CarRacingWebController.java:
```diff
@@ -0,0 +1,69 @@
+    @ExceptionHandler(InvalidFormatException.class)
+    public ResponseEntity<String> handleInvalidFormat(InvalidFormatException exception) {
+        final String exceptionMessage = "[ERROR] 시도할 횟수에 숫자를 입력할 수 없습니다. 입력 값 : " + exception.getValue();
+        log.error("Unexpected error occurred : {}", exceptionMessage, exception);
```
예외가 유실되지 않게 잘 적용해주셨네요! 시간이 남으신다면 로그레벨에 대해서도 본인만의 기준 또는 다른 사람들이 사용하는 기준에 대해 학습해보면 좋습니다😀

KIMSIYOUNG commented on src/main/java/racingcar/controller/CarRacingWebController.java:
```diff
@@ -0,0 +1,69 @@
+
+    @ExceptionHandler(InvalidFormatException.class)
+    public ResponseEntity<String> handleInvalidFormat(InvalidFormatException exception) {
+        final String exceptionMessage = "[ERROR] 시도할 횟수에 숫자를 입력할 수 없습니다. 입력 값 : " + exception.getValue();
```
프로그램내에서 발생하는 `InvalidFormatException`은 항상 이 동작에서만 일어난다는 보장이 있을까요?

KIMSIYOUNG commented on src/main/java/racingcar/controller/dto/RacingStartDTO.java:
```diff
@@ -0,0 +1,29 @@
+public class RacingStartDTO {
+
+    @NotBlank(message = "이름을 입력해야 합니다.")
+    private String names;
```
이 부분은 직접 이것저것 삽질해가며 배워가시면 좋을 것 같습니다. 
일단 서버에서는 아래와 같이 받으면 됩니다. 
프론트를 깊게 알 필욘 없지만 어떻게 요청을 쏴야하는지정도는 알고 있어야 좋을 것 같아 프론트 부분은 따로 가이드 드리지 않을게요!

```java
@NotEmpty
private List<@NotBlank String> name;

```

KIMSIYOUNG commented on src/main/java/racingcar/controller/CarRacingWebController.java:
```diff
@@ -0,0 +1,69 @@
+        this.carRacingService = carRacingService;
+    }
+
+    @PostMapping("/plays")
```
`@RequestMapping` 이라는 어노테이션도 있습니다. 참고만 해주세요!

KIMSIYOUNG commented on src/main/java/racingcar/controller/CarRacingWebController.java:
```diff
@@ -0,0 +1,69 @@
+        return ResponseEntity.ok(racingResultDTO);
+    }
+
+    @GetMapping("/plays")
```
여유가 되신다면 각 XXXMapping의 차이에 대해서도 학습해보시면 좋을 것 같습니다.

KIMSIYOUNG commented on src/main/java/racingcar/dao/LocalCarDao.java:
```diff
@@ -0,0 +1,87 @@
+
+    private static final int UPDATED_ROWS = 1;
+
+    private final Map<Long, CarEntity> carEntities = new LinkedHashMap<>();
```
`Static` 영역으로 구현되어야 할 것 같네요(밑의 아이디도 동일)

KIMSIYOUNG commented on src/main/java/racingcar/service/CarRacingService.java:
```diff
@@ -0,0 +1,97 @@
+
+        return gameIds.stream()
+                .map(gameIdDTO -> new RacingResultDTO(
+                        mapToWinners(carDao.findWinners(gameIdDTO.getId())),
```
위너를 저장하는 방식과 저장하지 않는 방식엔 어떤 차이와 장단이 있나요?

KIMSIYOUNG commented on src/test/java/racingcar/dao/LocalCarDaoTest.java:
```diff
@@ -0,0 +1,114 @@
+
+    @BeforeEach
+    void setUp() {
+        gameId = gameDao.insert(2);
```
테스트가 여러개가 실행되면 데이터가 중복되어 저장되어 테스트가 꺠질 수도 있습니다.
이 부분은 어떻게 해결하면 좋을지 고민해보면 좋을 것 같습니다.

KIMSIYOUNG commented on src/main/resources/data.sql:
```diff
@@ -15,5 +15,5 @@ CREATE TABLE CAR
-    FOREIGN KEY (game_id) REFERENCES GAME (id)
-);
\ No newline at end of file
+    FOREIGN KEY (game_id) REFERENCES GAME (id) ON DELETE CASCADE
```
포린키를 거는 것과 걸지 않는 것의 장단도 고민해보면 좋을 것 같아요.

KIMSIYOUNG reviewed (approved):
안녕하세요 후추
리뷰 잘 반영해주셨네요.
몇가지 코멘트 추가로 남겼으니 확인 부탁드립니다.
미션에서 필요한 내용은 이미 잘 구현해주신 것 같아 머지하겠습니다.
궁금하신 내용이나 추가로 이야기 나누고 싶은 내용이 있다면 디엠 부탁드릴게요!
